src/hooks/protections/receive_net_message.cpp

```diff
@@ -98,227 +321,439 @@ namespace big
 			}
 		}
 
-		if (!get_msg_type(msgType, buffer))
-			return g_hooking->get_original<hooks::receive_net_message>()(netConnectionManager, a2, frame);
+		int sec_id = 0;
 
+		if (frame)
+			sec_id = frame->m_security_id;
+		else
+		{
+			if (auto cxn = g_pointers->m_gta.m_get_connection_peer(net_cxn_mgr, error->m_peer_id))
+				sec_id = cxn->m_security_id;
+		}
 
-		if (msgType == rage::eNetMessage::MsgRoamingJoinBubbleAck)
+		auto peer = g_pointers->m_gta.m_get_peer_by_security_id(sec_id); // shouldn't be null in most cases, contains unspoofable data
+
+		if (error && msgType != rage::eNetMessage::MsgJoinResponse)
+		{
+			if (peer)
+				LOGF(stream::net_messages, WARNING, "Received an error packet that isn't MsgJoinResponse from {}", peer->m_info.name);
 			return true;
+		}
+
+		if (g.debug.logs.packet_logs)
+		{
+			log_net_message(msgType, buffer, event, peer);
+		}
 
-		if (player)
+		switch (msgType)
 		{
-			switch (msgType)
+		case rage::eNetMessage::MsgScriptJoin:
+		{
+			CGameScriptId script;
+			script_id_deserialize(script, buffer);
+
+			if (player && handle_block_script(player, script, event->m_msg_id))
 			{
-			case rage::eNetMessage::MsgTextMessage:
-			case rage::eNetMessage::MsgTextMessage2:
+				LOGF(stream::net_messages, WARNING, "Denying script request from {} (hash={:X}, instance={})", player->get_name(), script.m_hash, script.m_instance_id);
+				return true;
+			}
+
+			break;
+		}
+		case rage::eNetMessage::MsgScriptHostRequest:
+		{
+			CGameScriptId script;
+			script_id_deserialize(script, buffer);
+
+			if (script.m_hash == "freemode"_J && g.session.force_script_host)
+				return true;
+
+			break;
+		}
+		case rage::eNetMessage::MsgKickPlayer:
+		{
+			KickReason reason = buffer.Read<KickReason>(3);
+
+			if (!is_host_of_session(gta_util::get_network()->m_game_session_ptr, event->m_peer_id))
 			{
-				char message[256];
-				rage::rlGamerHandle handle{};
-				bool is_team;
-				buffer.ReadString(message, sizeof(message));
-				gamer_handle_deserialize(handle, buffer);
-				buffer.ReadBool(&is_team);
+				LOGF(stream::net_messages, WARNING, "{} sent MsgKickPlayer, but they are not the host", peer->m_info.name);
+				return true;
+			}
 
-				if (player->is_spammer)
+			if (reason == KickReason::VOTED_OUT)
+			{
+				g_notification_service.push_warning("PROTECTIONS"_T.data(), "YOU_HAVE_BEEN_KICKED"_T.data());
+				return true;
+			}
+
+			LOGF(stream::net_messages, VERBOSE, "{} sent us a MsgKickPlayer, reason = {}", peer->m_info.name, (int)reason);
+			break;
+		}
+		case rage::eNetMessage::MsgRadioStationSyncRequest:
+		{
+			if (player && player->block_radio_requests)
+				return true;
+
+			if (!session || !session->is_host())
+			{
+				LOGF(stream::net_messages, WARNING, "{} sent MsgRadioStationSyncRequest, but we are not the host", player->get_name());
```

Comment:
when some try to oom kick you and you are not host. This log will spam like 2000 times